lib/spack/spack/cmd/deconcretize.py

```diff
@@ -0,0 +1,96 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import sys
+
+import llnl.util.tty as tty
+
+import spack.cmd
+import spack.cmd.common.arguments as arguments
+import spack.cmd.common.confirmation as confirmation
+import spack.spec
+
+description = "remove specs from the concretized lockfile of an environment"
+section = "environments"
+level = "long"
+
+# Arguments for display_specs when we find ambiguity
+display_args = {"long": True, "show_flags": False, "variants": False, "indent": 4}
+
+
+def setup_parser(subparser):
+    subparser.add_argument(
+        "--root",
+        action="store_true",
+        help="deconcretize only environment roots. additional roots that depend on matching specs"
+        " will be ignored",
+    )
+    arguments.add_common_arguments(subparser, ["yes_to_all", "specs"])
+    subparser.add_argument(
+        "-a",
+        "--all",
+        action="store_true",
+        dest="all",
+        help="deconcretize ALL specs that match each supplied spec",
+    )
+
+
+def get_deconcretize_list(args, specs, env):
```

Comment:
doctoring should say something about this returning a list of *roots* to deconcretize.  Also type hints would be nice here... 

```suggestion
def get_deconcretize_list(
    args: argparse.Namespace,
    specs: List[spack.spec.Spec],
    env: Environment
) -> List[spack.spec.Spec]:
```